pennylane/devices/null_qubit.py

```diff
@@ -11,320 +11,392 @@
+logger.addHandler(logging.NullHandler())
+
+Result_or_ResultBatch = Union[Result, ResultBatch]
+QuantumTapeBatch = Sequence[QuantumTape]
+QuantumTape_or_Batch = Union[QuantumTape, QuantumTapeBatch]
+# always a function from a resultbatch to either a result or a result batch
+PostprocessingFn = Callable[[ResultBatch], Result_or_ResultBatch]
+
+
+@singledispatch
+def null_measurement(
+    mp: MeasurementProcess, obj_with_wires, shots: Shots, batch_size: int, interface: str
+):
+    """Create all-zero results for various measurement processes."""
+    return _null_measurement(mp, obj_with_wires, shots, batch_size, interface)
+
+
+def _null_measurement(mp, obj_with_wires, shots, batch_size, interface):
+    shape = mp.shape(obj_with_wires, shots)
+    if all(isinstance(s, int) for s in shape):
+        if batch_size is not None:
+            shape = (batch_size,) + shape
+        return math.zeros(shape, like=interface, dtype=mp.numeric_type)
+    if batch_size is not None:
+        shape = ((batch_size,) + s for s in shape)
+    return tuple(math.zeros(s, like=interface, dtype=mp.numeric_type) for s in shape)
+
+
+@null_measurement.register
+def _(mp: ClassicalShadowMP, obj_with_wires, shots, batch_size, interface):
+    shapes = [mp.shape(obj_with_wires, Shots(s)) for s in shots]
+    if batch_size is not None:
+        # shapes = [(batch_size,) + shape for shape in shapes]
+        raise ValueError(
+            "Parameter broadcasting is not supported with null.qubit and qml.classical_shadow"
+        )
+    results = tuple(math.zeros(shape, like=interface, dtype=np.int8) for shape in shapes)
+    return results if shots.has_partitioned_shots else results[0]
+
+
+@null_measurement.register
+def _(mp: CountsMP, obj_with_wires, shots, batch_size, interface):
+    outcomes = []
+    if mp.obs is None and not isinstance(mp.mv, MeasurementValue):
+        num_wires = len(obj_with_wires.wires)
+        state = "0" * num_wires
+        results = tuple({state: math.asarray(s, like=interface)} for s in shots)
+        if mp.all_outcomes:
+            outcomes = [f"{x:0{num_wires}b}" for x in range(1, 2**num_wires)]
+    else:
+        outcomes = sorted(mp.eigvals())  # always assign shots to the smallest
+        results = tuple({outcomes[0]: math.asarray(s, like=interface)} for s in shots)
+        outcomes = outcomes[1:] if mp.all_outcomes else []
+
+    if outcomes:
+        zero = math.asarray(0, like=interface)
+        for res in results:
+            for val in outcomes:
+                res[val] = zero
+    if batch_size is not None:
+        results = tuple([r] * batch_size for r in results)
+    return results[0] if len(results) == 1 else results
+
+
+@null_measurement.register(StateMP)
+@null_measurement.register(ProbabilityMP)
+def _(mp: Union[StateMP, ProbabilityMP], obj_with_wires, shots, batch_size, interface):
+    if isinstance(mp, DensityMatrixMP):
+        return _null_measurement(mp, obj_with_wires, shots, batch_size, interface)
```

Comment:
Why not just register an additional one for `DensityMatrixMP`?
```
null_measurement.register(DensityMatrixMP)(_null_measurement)
```

Reply:
cool, adding that